webapp/src/main/java/com/box/l10n/mojito/service/pullrun/PullRunAssetService.java

```diff
@@ -1,26 +1,172 @@
 
         return pullRunAssetRepository.save(pullRunAsset);
     }
+
+    @StopWatch
+    @Transactional
+    public void saveTextUnitVariants(PullRunAsset pullRunAsset, List<Long> tuvIds) {
+        Repository repository = pullRunAsset.getPullRun().getRepository();
+        meterRegistry.timer("PullRunAssetService.saveTextUnitVariants", Tags.of("repositoryId", Objects.toString(repository.getId())))
+                .record(() -> {
+                    Lists.partition(tuvIds, BATCH_SIZE).forEach(tuvIdsBatch -> saveTextUnitVariantsBatch(pullRunAsset, tuvIdsBatch));
+                });
+    }
+
+    @StopWatch
+    @Transactional
+    public void saveTextUnitVariantsPreparedStatement(PullRunAsset pullRunAsset, List<Long> tuvIds) {
+        Repository repository = pullRunAsset.getPullRun().getRepository();
+        JdbcTemplate jdbcTemplate = new JdbcTemplate(dataSource);
+        Instant now = Instant.now();
+        meterRegistry.timer("PullRunAssetService.saveTextUnitVariantsPreparedStatement", Tags.of("repositoryId", Objects.toString(repository.getId())))
+                .record(() -> {
+                    Lists.partition(tuvIds, BATCH_SIZE).forEach(tuvIdsBatch -> saveTextUnitVariantsPreparedStatementBatch(jdbcTemplate, pullRunAsset, tuvIdsBatch, now));
+                });
+    }
+
+    @StopWatch
+    @Transactional
+    public void saveTextUnitVariantsMultiRow(PullRunAsset pullRunAsset, List<Long> tuvIds) {
+        Repository repository = pullRunAsset.getPullRun().getRepository();
+        JdbcTemplate jdbcTemplate = new JdbcTemplate(dataSource);
+        Instant now = Instant.now();
+        meterRegistry.timer("PullRunAssetService.saveTextUnitVariantsMultiRow", Tags.of("repositoryId", Objects.toString(repository.getId())))
+                .record(() -> {
+                    Lists.partition(tuvIds, BATCH_SIZE).forEach(tuvIdsBatch -> saveTextUnitVariantsMultiRowBatch(jdbcTemplate, pullRunAsset, tuvIdsBatch, now));
+                });
+    }
+
+    /**
+     * TODO(jean) Other implementation will track translation inheritance out of the box, but not this.
+     * Though actually with current implementation the inheritance handling might be complicated. To review - how
+     * do you know fr-CA delta needs to be changed when fr-FR is changed?
+     * <p>
+     * Ran locally this was very slow.. while we expected it to be the faster option
+     */
+    @StopWatch
+    @Transactional
+    public void saveTextUnitVariantsServer(PullRunAsset pullRunAsset, List<Long> tuvIds, Long localeId) {
+        Repository repository = pullRunAsset.getPullRun().getRepository();
+        JdbcTemplate jdbcTemplate = new JdbcTemplate(dataSource);
+        meterRegistry.timer("PullRunAssetService.saveTextUnitVariantsMultiRow", Tags.of("repositoryId", Objects.toString(repository.getId())))
+                .record(() -> {
+                    // TODO(jean) might be the fastest option (if data is clean on the server side),
```

Comment:
@geoarion  on my localhost it turned out to be slow ... but everything was slow compared to my original test